Use new error type & log in session close

anagha-ks opened the pull request:
<!-- Provide a general summary of your changes in the title above

Please read our contribution guidelines and sign the Contributor License
Agreement (CLA) before submitting the pull request. Also, check if there are no
other open pull requests targeting the same issue. -->

#### Description
For CloseSession API:

- Update it to return the newly defined error type (#132)
- Log the request, error/success response are logged in a structured
format (#133)

##### Category
<!-- Tell us what type of issue does your pull request target.
You can uncomment one of the following options: -->

<!-- Bug Fix -->
<!-- Improvement -->
Implementation Task

##### Relevant issue
Relates to #132, #133

#### Testing
All tests should pass

##### Steps to run the tests
<!-- Describe a set of steps to run the tests relevant to this change. -->

1. 
2. 
3.  

#### Checklist 
<!-- Please check if the pull request fulfils these requirements: -->

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

manoranjith commented on session/session.go:
```diff
@@ -62,12 +62,14 @@ func (e Error) Error() string {
+	ErrNoActiveSub          Error = "no active subscription"
+	ErrChClosed             Error = "Channel is closed"
+	ErrUnknownChID          Error = "No channel corresponding to the specified ID"
+	ErrunexpectedPhaseChIDs Error = "channels in unexpected phase during session close"
```
```suggestion
```
This one need not be defined as a named error. You can return it as unknown internal error.

manoranjith commented on session/session.go:
```diff
@@ -772,20 +777,27 @@ func (s *Session) Close(force bool) ([]perun.ChInfo, error) {
-		err := fmt.Errorf("chs in unexpected phase during session close: %v", unexpectedPhaseChIDs)
-		s.Error(err.Error())
+		err := ErrunexpectedPhaseChIDs
+		apiErr = perun.NewAPIErrV2FailedPreCondition(err.Error())
```
In a proper scenario, this error should not occur. So you can mark this as unknown internal error.

manoranjith commented on session/session.go:
```diff
@@ -772,20 +777,27 @@ func (s *Session) Close(force bool) ([]perun.ChInfo, error) {
 	if !force && len(openChsInfo) != 0 {
-		err := fmt.Errorf("%w: %v", perun.ErrOpenCh, openChsInfo)
-		s.Error(err.Error())
+		apiErr = perun.NewAPIErrV2FailedPreCondition(ErrOpenCh.Error())
```
Can you update the `NewAPIErrV2FailedPreconditio` to take `addInfo` as an argument. And in this case, you can create an addInfo type with the a field containing open channels info ?

manoranjith commented on session/session.go:
```diff
@@ -772,20 +777,27 @@ func (s *Session) Close(force bool) ([]perun.ChInfo, error) {
 
 	s.isOpen = false
-	return openChsInfo, s.close()
+	err := s.close()
```
I think, this err does not have any precondition. Should this error instead be treated as as unknown internal error ?

manoranjith commented on api/grpc/server_integ_test.go:
```diff
@@ -159,15 +159,6 @@ func Test_Integ_Role(t *testing.T) {
 	})
 
-	t.Run("CloseSession_NoForce_ErrOpenPayChs", func(t *testing.T) {
-		openPayChsInfo := CloseSession(t, aliceSessionID, false, true)
```
Why should we remove this test case ?

manoranjith commented on api/grpc/pb/api.proto:
```diff
@@ -298,15 +298,11 @@ message CloseSessionReq {
     oneof response {
         MsgSuccess msgSuccess = 1;
-        MsgError error = 2;
+        MsgErrorV2 error = 2;
```
Should we also add `ErrV2InfoFailedPreCondition` in line 63 (above) ?

manoranjith commented on api/grpc/server.go:
```diff
@@ -697,6 +714,13 @@ func toGrpcError(err perun.APIErrorV2) *pb.MsgErrorV2 {
+
+		grpcErr.AddInfo = &pb.MsgErrorV2_ErrV2InfoFailedPreConditionSessionClose{
+			ErrV2InfoFailedPreConditionSessionClose: &pb.ErrV2InfoFailedPreConditionSessionClose{
+				ChInfo: toGrpcChsInfo(info.OpenChsInfo),
```
We should be returning a "payment channels info" instead of "channels info". To do this, you will have to convert the "channels info" in the `error.AddInfo` field to "payment channels info" in the `app/payment` package.

After that, in `grpc` wrapper, you will be having "payment channels info" which you can process using the already existing `toGrpcPayChsInfo` function.

manoranjith commented on api/grpc/server.go:
```diff
@@ -666,6 +663,26 @@ func toGrpcBalInfo(src perun.BalInfo) *pb.BalInfo {
 	}
 }
 
+// toGrpcChsInfo is a helper function to convert slice of ChInfo struct defined in perun-node
```
Once you implement the below comment, these functions will not be needed. So you can remove these and use `toGrpcPayChsInfo`which is already defined.

manoranjith commented on session/session.go:
```diff
@@ -774,19 +777,29 @@ func (s *Session) Close(force bool) ([]perun.ChInfo, error) {
+			OpenChsInfo: openChsInfo,
+		}
+		apiErr = perun.NewAPIErrV2FailedPreCondition(openCh, ErrOpenCh.Error())
+		s.WithFields(perun.APIErrV2AsMap("Close", apiErr)).Error(apiErr.Message())
```
```suggestion
		addInfo := perun.ErrV2InfoFailedPreConditionSessionClose{
			OpenChsInfo: openChsInfo,
		}
		apiErr = perun.NewAPIErrV2FailedPreCondition(addInfo, ErrOpenCh.Error())
		s.WithFields(perun.APIErrV2AsMap("Close", apiErr)).Error(apiErr.Message())
```

manoranjith commented on app/payment/session.go:
```diff
@@ -90,7 +90,7 @@ func RespondPayChProposal(pctx context.Context, s perun.SessionAPI, proposalID s
 
 // CloseSession closes the current session.
-func CloseSession(s perun.SessionAPI, force bool) ([]PayChInfo, error) {
+func CloseSession(s perun.SessionAPI, force bool) ([]PayChInfo, perun.APIErrorV2) {
```
You should also do type assertion on the error message. If it is "Failed Precondition Error" with open channels, then the channels info should be converted to payment channels info.

manoranjith commented on session/session_test.go:
```diff
@@ -695,40 +695,44 @@ func Test_ProposeCh_CloseSession(t *testing.T) {
 		require.Error(t, err)
-		t.Log(err)
+
+		wantMessage := session.ErrOpenCh.Error()
```
Can you rebase this PR on top of #150 ? There, I have added some assert helper functions (to the end of `session/session_test.go` file). Using those helper functions, will make these assert statements to one liners that are easy to read.

Commit:
Use new error type & log in session Close

For Session Close API:
- Update it to return the newly defined error type (#132).

- Log the request, error/success responses in a structured format.
  (#133).

Relates to #132, #133.

Co-authored-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>
Signed-off-by: Manoranjith <ponraj.manoranjitha@in.bosch.com>

danielksan81 reviewed (approved):
ok
